.circleci/config.yml

```diff
@@ -1,91 +1,145 @@
+            echo 'export TF_VAR_tags=\{\"CIRCLE_BUILD_URL\"=\"${CIRCLE_BUILD_URL}\",\"CIRCLE_PR_NUMBER\"=\"${CIRCLE_PR_NUMBER}\",\"CIRCLE_PULL_REQUEST\"=\"${CIRCLE_PULL_REQUEST}\",\"CIRCLE_REPOSITORY_URL\"=\"${CIRCLE_REPOSITORY_URL}\"\}' >> $BASH_ENV
+  install_tf:
     parameters:
-      workspace:
-        type: string
-        default: circleci-<< pipeline.number >>
       terraform_version:
         type: string
-    environment:
-      WORKSPACE: << parameters.workspace >>
-
+        default: << pipeline.parameters.terraform_version >>
     steps:
-      - checkout
-
       - terraform/install:
-          terraform_version: <<parameters.terraform_version>>
-
+          terraform_version: << parameters.terraform_version >>
+  gen_pvt_key:
+    steps:
       - run:
-          name: Configure test cluster env
-          command: |
-            echo export TF_VAR_deploy_id=\"${WORKSPACE}\" >> $BASH_ENV
-            echo 'export TF_VAR_tags=\{\"CIRCLE_BUILD_URL\"=\"${CIRCLE_BUILD_URL}\",\"CIRCLE_PR_NUMBER\"=\"${CIRCLE_PR_NUMBER}\",\"CIRCLE_PULL_REQUEST\"=\"${CIRCLE_PULL_REQUEST}\",\"CIRCLE_REPOSITORY_URL\"=\"${CIRCLE_REPOSITORY_URL}\"\}' >> $BASH_ENV
-
+          name: Generate pvt key
+          command: ssh-keygen -q -P '' -t rsa -b 4096 -m PEM -f tests/domino.pem
+  tf_init_apply:
+    steps:
       - run:
-          name: Test Terraform
+          name: Terraform init/validate/apply
           working_directory: tests
           command: |
-            ssh-keygen -q -P '' -t rsa -b 4096 -m PEM -f domino.pem
+            echo "Current dir: $(pwd)"
             terraform init
             terraform validate
             terraform workspace new ${WORKSPACE}
             terraform apply -auto-approve
-            # we will always have drift
-            # terraform plan -detailed-exitcode
-
+  tf_destroy:
+    steps:
       - run:
           name: Terraform destroy
           working_directory: tests
           command: |
-            terraform destroy -auto-approve
+            echo "Current dir: $(pwd)"
+            terraform destroy --auto-approve || terraform destroy --auto-approve --refresh=false
           when: always
-
+  tf_ws_delete:
+    steps:
       - run:
           name: Terraform workspace delete
           working_directory: tests
           command: |
+            echo "Current dir: $(pwd)"
             terraform workspace select default
             terraform workspace delete ${WORKSPACE}
+          when: always
 
-parameters:
-  terraform_version:
-    default: "1.3.6"
-    type: string
+jobs:
+  test-deploy:
+    docker:
+      - image: cimg/aws:2022.11.1
+    parameters:
+      workspace:
+        type: string
+        default: circleci-<< pipeline.number >>
+      terraform_version:
+        type: string
+    environment:
+      WORKSPACE: << parameters.workspace >>
+    steps:
+      - checkout
+      - install_tf
+      - set_env_vars
+      - gen_pvt_key
+      - tf_init_apply
+      - tf_destroy
+      - tf_ws_delete
```

Comment:
shouldn't these have a when: always?

Reply:
[here](https://github.com/dominodatalab/terraform-aws-eks/blob/235b3b573b300de75b9292b5f95b3272e5f0011e/.circleci/config.yml#L60) and [here](https://github.com/dominodatalab/terraform-aws-eks/blob/235b3b573b300de75b9292b5f95b3272e5f0011e/.circleci/config.yml#L70)

Reply:
what is unclear is whether a command as a job step is implicity replace or just treated as a separate sequence of steps. In which case the job steps above would need to denote that a given command will always execute and most likely the steps within the command as well.
The CircleCI docs are unclear on the exact semantics. But if you have seen errors from any previous step cause destroy and delete, I am ok with it.

Reply:
The docs claim the command steps are inserted so what you have will work.